src/Databases/DatabaseOnDisk.cpp

```diff
@@ -751,8 +753,57 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromStorage(const String & table_name, cons
     return create_table_query;
 }
 
-void DatabaseOnDisk::modifySettingsMetadata(const SettingsChanges & settings_changes, ContextPtr query_context)
+void DatabaseOnDisk::persistCreateDatabaseQuery(const ASTPtr & ast, const DatabaseMetadataPersistedCallback & commit_cb) const
+{
+    std::lock_guard lock(mutex);
+    persistCreateDatabaseQueryUnlocked(ast, commit_cb);
+}
+
+void DatabaseOnDisk::persistCreateDatabaseQueryUnlocked(const ASTPtr & ast, const DatabaseMetadataPersistedCallback & commit_cb) const
+{
+    auto database_metadata_path = getDatabaseMetadataFilePath();
+    ASTPtr ast_clone = ast->clone();
+    auto & create = ast_clone->as<ASTCreateQuery &>();
+    if (create.storage && create.storage->engine && create.storage->engine->name != getEngineName())
+    {
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Altering database engine is not allowed");
+    }
```

Comment:
seems like this check should be on a different level, before we even start to do `persistCreateDatabaseQueryUnlocked`